src/librustc/traits/error_reporting.rs

```diff
@@ -730,40 +730,35 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                                                    expected_found.found,
                                                    expected_trait_ty.is_closure())
                 } else if let &TypeError::Sorts(ref expected_found) = e {
-                    let expected = if let ty::TyTuple(tys, _) = expected_found.expected.sty {
```

Comment:
You're still matching against the expected/found, which could lead to this bogusly triggering when one of the *arguments* is a tuple - when you have something like `fn((u32, u32)): Fn<(u32,)>`.

There's no reason to do so - you have the trait-refs in `expected_trait_ref` and `actual_trait_ref`.

You can get the argument length of each argument in this way:
```Rust
// skipping the binder is ok because we only care if this is a tuple
let arg_len = match X_trait_ref.skip_binder().substs.type_at(1) {
    ty::TyTuple(ref tys) => Some(tys.len()),
    _ => None
};
```
```